File: 496-next-greater-element-i/next-greater-element-i.java
import java.util.*;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {

        Stack<Integer> stack = new Stack<>();
        HashMap<Integer, Integer> map = new HashMap<>();

        // process nums2
        for (int i = 0; i < nums2.length; i++) {

            while (!stack.isEmpty() && stack.peek() < nums2[i]) {
                map.put(stack.pop(), nums2[i]);
            }

            stack.push(nums2[i]);
        }

        // build answer (use -1 if not found in map)
        int[] ans = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            ans[i] = map.getOrDefault(nums1[i], -1);
        }

        return ans;
    }
}

File: monotonic-stack/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> st;
        vector<int> greater(nums2.size());
        for(int i=nums2.size()-1;i>=0;i--){
            while(!st.empty() && nums2[i]>=st.top()){
                st.pop();
            }

            if(st.empty()) greater[i] = -1;
            else greater[i] = st.top();

            st.push(nums2[i]);
        }

        for(int i=0;i<nums1.size();i++){
            auto idx = find(nums2.begin(),nums2.end(),nums1[i]);
            int id = idx - nums2.begin();
            nums1[i] = greater[id];
        }
        return nums1;
    }
};